src/scipp/core/binning.py

```diff
@@ -0,0 +1,528 @@
+    dimension labels.
+
+    When histogramming by non-dimension-coords, the output will have new dimensions
+    given by the names of these coordinates. These new dimensions replace the
+    dimensions the input coordinates depend on.
+
+    Parameters
+    ----------
+    x:
+        Input data.
+    arg_dict:
+        Dictionary mapping dimension labels to binning parameters.
+    **kwargs:
+        Mapping of dimension label to corresponding binning parameters.
+
+    Returns
+    -------
+    :
+        Histogrammed data.
+
+    See Also
+    --------
+    scipp.bin:
+        Creating binned data by binning instead of summing all contributions.
+    scipp.binning.make_histogrammed:
+        Lower level function for histogramming that does not automatically
+        replace/erase dimensions.
+
+    Examples
+    --------
+
+    Histogram a table by one of its coord columns, specifying (1) number of bins, (2)
+    bin width, or (3) actual binning:
+
+      >>> table = sc.data.table_xyz(100)
+      >>> table.hist(x=2).sizes
+      {'x': 2}
+
+      >>> table.hist(x=sc.scalar(0.2, unit='m')).sizes
+      {'x': 5}
+
+      >>> table.hist(x=sc.linspace('x', 0.2, 0.8, num=10, unit='m')).sizes
+      {'x': 9}
+
+    Histogram a table by two of its coord columns:
+
+      >>> table = sc.data.table_xyz(100)
+      >>> table.hist(x=4, y=6).sizes
+      {'x': 4, 'y': 6}
+
+    Histogram binned data, using existing bins:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> binned.hist().sizes
+      {'x': 10}
+
+    Histogram binned data, using new bins along existing dimension:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> binned.hist(x=20).sizes
+      {'x': 20}
+
+    Histogram binned data along an additional dimension:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> binned.hist(y=5).sizes
+      {'x': 10, 'y': 5}
+    """
+    edges = _make_edges(x, arg_dict, kwargs)
+    erase = _find_replaced_dims(x, edges)
+    if len(edges) == 0:
+        if x.bins is None:
+            raise TypeError("Data is not binned so bin edges must be provided.")
+        return x.bins.sum()
+    if len(edges) == 1:
+        # TODO Note that this may swap dims, is that ok?
+        out = make_histogrammed(x, bins=list(edges.values())[0])
+    else:
+        edges = list(edges.values())
+        out = make_histogrammed(make_binned(x, edges=edges[:-1], erase=erase),
+                                bins=edges[-1])
+    for dim in erase:
+        if dim in out.dims:
+            out = out.sum(dim)
+    return out
+
+
+def nanhist(x: Union[_cpp.DataArray, _cpp.Dataset],
+            arg_dict: Optional[Dict[str, Union[int, Variable]]] = None,
+            /,
+            **kwargs: Union[int, Variable]) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    """Compute a histogram, skipping NaN values.
+
+    Like :py:func:`scipp.hist`, but NaN values are skipped. See there for details and
+    examples.
+    """
```

Comment:
It is still worth copy-pasting the params and returns sections just so that they are rendered properly.